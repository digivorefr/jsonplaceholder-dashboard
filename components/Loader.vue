<template>
  <div class="loader" />
</template>

<script lang="
    ts"
  >
import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentLoader'
})
</script>
  </div>
</template>
