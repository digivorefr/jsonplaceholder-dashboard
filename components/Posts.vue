<template>
  <div :class="`user__posts${($fetchState.error !== null) ? ' error' : ''}`">
    <h2>Posts</h2>
    <h3
      v-if="$fetchState.error !== null"
      class="error__title"
    >
      An error occured.
    </h3>
    <Loader v-else-if="$fetchState.pending === true" />
    <h3 v-else-if="posts.length === 0">
      No posts
    </h3>
    <div
      v-else
      class="user__posts__wrapper"
    >
      <div class="user__posts__grid">
        <div
          v-for="post in posts"
          :key="post.id"
          class="user__post"
        >
          <h4 class="user__post__title">
            {{ post.title }}
          </h4>
          <p class="user__post__content">
            {{ post.body }}
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import Loader from '~/components/Loader.vue'

export type Post = {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export default Vue.extend({
  name: 'UserPosts',
  components: { Loader },
  props: {
    id: {
      type: Number,
      required: true
    }
  },
  data() {
    return {
      posts: [] as Post[]
    }
  },
  async fetch() {
    this.posts = await this.getPostsByUserId()
  },
  methods: {
    async getPostsByUserId(): Promise<Post[]> {
      return await this.$axios.$get(`/posts?userId=${this.id}`)
    }
  }
})
</script>
