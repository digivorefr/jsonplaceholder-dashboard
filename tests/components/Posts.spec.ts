import { mount, Wrapper } from '@vue/test-utils'
import Posts from '@/components/Posts.vue'
import Loader from '@/components/Loader.vue'

const postsMock = [
  {
    userId: 1,
    id: 1,
    title: 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
    body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'
  },
  {
    userId: 1,
    id: 2,
    title: 'qui est esse',
    body: 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'
  }
]

describe('Posts', () => {
  test('renders correctly', () => {
    const wrapper = mount(Posts, {
      mocks: {
        $fetchState: { pending: true, error: true, timestamp: Date.now() }
      }
    })
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.exists()).toBe(true)
    wrapper.destroy()
  })
  test('show error if fetch fails', () => {
    const wrapper = mount(Posts, {
      mocks: {
        $fetchState: { pending: null, error: true, timestamp: Date.now() }
      }
    })
    expect(wrapper.html()).toContain('An error occured.')
    wrapper.destroy()
  })
  test('show loader while fetching', () => {
    const wrapper = mount(Posts, {
      mocks: {
        $fetchState: { pending: true, error: null, timestamp: Date.now() }
      }
    })
    expect(wrapper.findComponent(Loader))
    wrapper.destroy()
  })
  test('show no posts text', () => {
    const wrapper = mount(Posts, {
      data() {
        return { posts: [] }
      },
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() }
      }
    })
    expect(wrapper.html()).toContain('No posts')
    wrapper.destroy()
  })
  test('Fetch and display posts', async () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = mount(Posts, {
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() },
        $axios: { $get: () => postsMock }
      }
    })
    wrapper.vm.$data.posts = await wrapper.vm.getPostsByUserId()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.$data.posts.length).toBe(2)
    expect(wrapper.findAll('.user__post').length).toBe(2)
    expect(wrapper.element).toMatchSnapshot()
    wrapper.destroy()
  })
})
