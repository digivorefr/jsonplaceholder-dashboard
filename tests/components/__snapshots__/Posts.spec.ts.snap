// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts Fetch and display posts 1`] = `
<div
  class="user__posts"
>
  <h2>
    Posts
  </h2>
   
  <div
    class="user__posts__wrapper"
  >
    <div
      class="user__posts__grid"
    >
      <div
        class="user__post"
      >
        <h4
          class="user__post__title"
        >
          
          sunt aut facere repellat provident occaecati excepturi optio reprehenderit
        
        </h4>
         
        <p
          class="user__post__content"
        >
          
          quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
        
        </p>
      </div>
      <div
        class="user__post"
      >
        <h4
          class="user__post__title"
        >
          
          qui est esse
        
        </h4>
         
        <p
          class="user__post__content"
        >
          
          est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
        
        </p>
      </div>
    </div>
  </div>
</div>
`;
